### PR TITLE
Rocket front-end can now fetch 4 instructions; added assert to dcache; refactoring

### PR DESCRIPTION
**Pull request by ccelio**
* The most important change is Rocket's front-end can now fetch 4-wide.
* Added assert() to catch if the core doesn't kill a cache request when a memory exception occurs
* renamed PCR to CSR
* renamed dmem.sret to invalidate_lr
* removed unused mbadaddr_wen in the CSR I/O

### CODE REVIEW

File: src/main/scala/icache.scala [hunk 1]
```diff
@@ -114,8 +114,13 @@ class Frontend(btb_updates_out_of_order: Boolean = false) extends FrontendModule
...
 
-
-  val fetch_data = icache.io.resp.bits.datablock >> (s2_pc(log2Up(rowBytes)-1,log2Up(coreFetchWidth*coreInstBytes)) << log2Up(coreFetchWidth*coreInstBits))
+  var fetch_data:Bits = null
+  require (coreFetchWidth <= 4)
+  if (coreFetchWidth == 4) {
+    fetch_data = icache.io.resp.bits.datablock
+  } else {
+    fetch_data = icache.io.resp.bits.datablock >> (s2_pc(log2Up(rowBytes)-1,log2Up(coreFetchWidth*coreInstBytes)) << log2Up(coreFetchWidth*coreInstBits))
+  }
```
Comment (aswaterman) [inline src/main/scala/icache.scala]: The magic constant 4 is super hacky and will break if the memory width changes.  The number derives from rowBytes / coreInstBytes.  Can't we do something like

require(coreFetchWidth * coreInstBytes <= rowBytes)
val fetch_data =
  if (coreFetchWidth * coreInstBytes == rowBytes) icache.io.resp.bits.datablock
  else icache.io.resp.bits.datablock >> (s2_pc(log2Up(rowBytes)-1,log2Up(coreFetchWidth*coreInstBytes)) << log2Up(coreFetchWidth*coreInstBits))

This also gets rid of the var.

**Comment by aswaterman** [comment]
This looks good if we address the line note I left on the I$ change.

**Comment by ccelio** [comment]
Thanks, that's good feedback. I addressed it with a new commit.

**Comment by aswaterman** [comment]
Looks good.  Thanks for the help.

**Pull request merged**